src/impl/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp

```diff
@@ -151,46 +173,70 @@ namespace Impl {
+    using XVector  = Kokkos::View<SCALAR const*, LAYOUT,device_type, Kokkos::MemoryTraits<Kokkos::Unmanaged|Kokkos::RandomAccess>>; \
+    using YVector  = Kokkos::View<SCALAR*, LAYOUT, device_type, memory_trait_type>; \
+    using Controls = KokkosKernels::Experimental::Controls;		\
 									\
     using coefficient_type = typename YVector::non_const_value_type;	\
 									\
-    static void spmv (const char mode[],				\
+    static void spmv (Controls& controls,				\
+                      const char mode[],				\
 		      const coefficient_type& alpha,			\
 		      const AMatrix& A,					\
 		      const XVector& x,					\
 		      const coefficient_type& beta,			\
 		      const YVector& y) {				\
-      std::string label = "KokkosSparse::spmv[TPL_CUSPARSE," + Kokkos::ArithTraits<SCALAR>::name() + "]"; \
-      Kokkos::Profiling::pushRegion(label);				\
-      spmv_cusparse(mode, alpha, A, x, beta, y);			\
-      Kokkos::Profiling::popRegion();					\
+      if(controls.getParameter("algorithm") == "native") {		\
+	std::string label = "KokkosSparse::spmv[TPL_CUSPARSE," + Kokkos::ArithTraits<SCALAR>::name() + "]"; \
+	Kokkos::Profiling::pushRegion(label);				\
+	spmv_native(controls, mode, alpha, A, x, beta, y);		\
+	Kokkos::Profiling::popRegion();					\
+      } else {								\
+	std::string label = "KokkosSparse::spmv[TPL_CUSPARSE," + Kokkos::ArithTraits<SCALAR>::name() + "]"; \
+	Kokkos::Profiling::pushRegion(label);				\
+	spmv_cusparse(controls, mode, alpha, A, x, beta, y);		\
+	Kokkos::Profiling::popRegion();					\
+      }									\
     }									\
   };
 
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, Kokkos::LayoutLeft,  true)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, Kokkos::LayoutRight, true)
-  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, Kokkos::LayoutRight, false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, Kokkos::LayoutLeft,  true)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, Kokkos::LayoutLeft,  false)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, Kokkos::LayoutRight, true)
-  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, Kokkos::LayoutRight, false)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutLeft,  true)
+  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutLeft,  false)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutRight, true)
+  // KOKKOSSPARSE_SPMV_CUSPARSE(double, int, int, Kokkos::LayoutRight, false)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutLeft,  true)
+  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutLeft,  false)
+  KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutRight, true)
+  // KOKKOSSPARSE_SPMV_CUSPARSE(float,  int, int, Kokkos::LayoutRight, false)
+#if defined(CUSPARSE_VERSION) && (10300 <= CUSPARSE_VERSION)
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int64_t, int, Kokkos::LayoutLeft,  true)
```

Comment:
I guess if we added the complex cudaDataTypes above, we could instantiate all of these for complex?

Reply:
Yes it should, I have not had time to look into complex yet though.
If in your experience it is easy to just instantiate and have it work OK through cusparse then I can do it. If it is an unknown I would suggest holding-off on it and do it in a later PR.